src/libraries/Common/src/System/Threading/AsyncOverSyncWithIoCancellation.cs

```diff
@@ -16,25 +15,58 @@ namespace System.Threading
+        /// <remarks>
+        /// The safety of caching this on the current thread is based on the fact that all work performed
+        /// is synchronous on this thread.  If a cancellation callback occurs on a different thread, that
+        /// could be using this instance, but only between the time that the operation is initiated on this
+        /// thread and disposal completes on this thread.
+        /// </remarks>
         [ThreadStatic]
-        private static SafeThreadHandle? t_currentThreadHandle;
+        private static AsyncOverSyncWithIoCancellation? t_instance;
 
         /// <summary>The OS handle of the thread performing the I/O.</summary>
-        private SafeThreadHandle? ThreadHandle;
+        /// <remarks>This is stored as part of the object's construction because the objects are thread affinitized.</remarks>
+        private readonly SafeThreadHandle? _threadHandle;
+
         /// <summary>Whether the call to CancellationToken.UnsafeRegister completed.</summary>
-        private volatile bool FinishedCancellationRegistration;
+        private bool _finishedCancellationRegistration;
         /// <summary>Whether the I/O operation has finished (successfully or unsuccessfully) and is requesting cancellation attempts stop.</summary>
-        private volatile bool ContinueTryingToCancel = true;
+        private bool _continueTryingToCancel;
         /// <summary>
         /// A task that may be checked after the <see cref="CancellationTokenRegistration"/> has been disposed.  If it's null at that point,
         /// the callback wasn't and will never be invoked.  If it's non-null, its completion represents the completion of the asynchronous callback.
         /// </summary>
-        private volatile Task? CallbackCompleted;
+        private Task? _callbackCompleted;
+
+        /// <summary>Initialize the instance.  This should be done once per thread.</summary>
+        private AsyncOverSyncWithIoCancellation()
+        {
+            // Get a handle for the current thread. This is stored and used to cancel the I/O on this thread
+            // in response to the cancellation token having cancellation requested.  If the handle is invalid,
+            // which could happen if OpenThread fails, skip attempts at cancellation. The handle needs to be
+            // opened with THREAD_TERMINATE in order to be able to call CancelSynchronousIo.
+            SafeThreadHandle handle = Interop.Kernel32.OpenThread(Interop.Kernel32.THREAD_TERMINATE, bInheritHandle: false, Interop.Kernel32.GetCurrentThreadId());
+            if (!handle.IsInvalid)
+            {
+                _threadHandle = handle;
+            }
+#if DEBUG
+            else
+            {
+                int lastError = Marshal.GetLastPInvokeError();
+                Debug.Fail($"{nameof(Interop.Kernel32.OpenThread)} unexpectedly failed with 0x{lastError:X8}: {Marshal.GetPInvokeErrorMessage(lastError)}");
+            }
+#endif
```

Comment:
Should we `Dispose` the handle in such case, so it does not need to wait for finalization? I imagine this is going to be very rare or even not happen at all. So nothing blocking here.

```suggestion
            else
            {
                handle.Dispose();
#if DEBUG
                int lastError = Marshal.GetLastPInvokeError();
                Debug.Fail($"{nameof(Interop.Kernel32.OpenThread)} unexpectedly failed with 0x{lastError:X8}: {Marshal.GetPInvokeErrorMessage(lastError)}");
#endif
            }
```